// ant-design-vue/es/vc-select/examples/single.js
import { defineComponent, createTextVNode, createVNode } from "vue";

/* eslint-disable no-console */
import Select, { Option } from '..';
import '../assets/index.less';
import './single.less';
var Test = defineComponent({
  data: function data() {
    return {
      destroy: false,
      value: '9'
    };
  },
  methods: {
    onChange: function onChange(e) {
      var value;

      if (e && e.target) {
        value = e.target.value;
      } else {
        value = e;
      }

      console.log('onChange', value);
      this.value = value;
    },
    onDestroy: function onDestroy() {
      this.destroy = 1;
    },
    onBlur: function onBlur(v) {
      console.log('onBlur', v);
    },
    onFocus: function onFocus() {
      console.log('onFocus');
    },
    onSearch: function onSearch(val) {
      console.log('Search:', val);
    }
  },
  render: function render() {
    var value = this.value,
        destroy = this.destroy;

    if (destroy) {
      return null;
    }

    return createVNode("div", {
      "style": {
        margin: '20px'
      }
    }, [createVNode("div", {
      "style": {
        height: '150px',
        background: 'rgba(0, 255, 0, 0.1)'
      },
      "onMousedown": function onMousedown(e) {
        e.preventDefault();
      }
    }, [createTextVNode("Prevent Default")]), createVNode("h2", null, [createTextVNode("Single Select")]), createVNode("div", {
      "style": {
        width: '300px'
      }
    }, [createVNode(Select, {
      "autofocus": true,
      "id": "my-select",
      "value": value,
      "placeholder": "placeholder",
      "showSearch": true,
      "style": {
        width: '500px'
      },
      "onBlur": this.onBlur,
      "onFocus": this.onFocus,
      "onSearch": this.onSearch,
      "allowClear": true,
      "optionFilterProp": "text",
      "onChange": this.onChange,
      "onPopupScroll": function onPopupScroll() {
        console.log('Scroll!');
      }
    }, {
      default: function _default() {
        return [createVNode(Option, {
          "value": null
        }, {
          default: function _default() {
            return [createTextVNode("\u4E0D\u9009\u62E9")];
          }
        }), createTextVNode(","), createVNode(Option, {
          "value": "01",
          "text": "jack",
          "title": "jack"
        }, {
          default: function _default() {
            return [createVNode("b", {
              "style": {
                color: 'red'
              }
            }, [createTextVNode("jack")])];
          }
        }), createTextVNode(","), createVNode(Option, {
          "value": "11",
          "text": "lucy"
        }, {
          default: function _default() {
            return [createVNode("span", null, [createTextVNode("lucy")])];
          }
        }), createTextVNode(","), createVNode(Option, {
          "value": "21",
          "disabled": true,
          "text": "disabled"
        }, {
          default: function _default() {
            return [createTextVNode("disabled")];
          }
        }), createTextVNode(","), createVNode(Option, {
          "value": "31",
          "text": "yiminghe",
          "class": "test-option",
          "style": {
            background: 'yellow'
          }
        }, {
          default: function _default() {
            return [createTextVNode("yiminghe")];
          }
        }), createTextVNode(","), [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map(function (i) {
          return createVNode(Option, {
            "key": i,
            "value": String(i),
            "text": String(i)
          }, {
            default: function _default() {
              return [i, createTextVNode("-text")];
            }
          });
        })];
      }
    })]), createVNode("h2", null, [createTextVNode("native select")]), createVNode("select", {
      "value": value,
      "style": {
        width: '500px'
      },
      "onChange": this.onChange
    }, [createVNode("option", {
      "value": "01"
    }, [createTextVNode("jack")]), createVNode("option", {
      "value": "11"
    }, [createTextVNode("lucy")]), createVNode("option", {
      "value": "21",
      "disabled": true
    }, [createTextVNode("disabled")]), createVNode("option", {
      "value": "31"
    }, [createTextVNode("yiminghe")]), [1, 2, 3, 4, 5, 6, 7, 8, 9].map(function (i) {
      return createVNode("option", {
        "value": i,
        "key": i
      }, [i]);
    })]), createVNode("h2", null, [createTextVNode("RTL Select")]), createVNode("div", {
      "style": {
        width: '300px'
      }
    }, [createVNode(Select, {
      "id": "my-select-rtl",
      "placeholder": "rtl",
      "direction": "rtl",
      "dropdownMatchSelectWidth": 300,
      "dropdownStyle": {
        minWidth: '300px'
      },
      "style": {
        width: '500px'
      }
    }, {
      default: function _default() {
        return [createVNode(Option, {
          "value": "1"
        }, {
          default: function _default() {
            return [createTextVNode("1")];
          }
        }), createVNode(Option, {
          "value": "2"
        }, {
          default: function _default() {
            return [createTextVNode("2")];
          }
        })];
      }
    })]), createVNode("p", null, [createVNode("button", {
      "type": "button",
      "onClick": this.onDestroy
    }, [createTextVNode("destroy")])])]);
  }
});
export default Test;
/* eslint-enable */
